File: src/lab3/StringType.java
package lab3;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;

public class StringType implements UserType<String> {

    @Override
    public String typeName() {
        return String.class.getTypeName();
    }

    @Override
    public String create() {
        return "";
    }

    @Override
    public String clone(String s) {
        return s;
    }

    @Override
    public String readValue(InputStreamReader in) throws IOException {
        try (BufferedReader br = new BufferedReader(in)) {
            String line = br.readLine();
            return line == null ? "" : line.trim();
        }
    }

    @Override
    public String parseValue(String stringVal) {
        return stringVal != null ? stringVal : "";
    }

    @Override
    public Comparator<String> getTypeComparator() {
        return String::compareTo;
    }
}

File: src/lab6kotlin/StringType.kt
package lab6kotlin

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class StringType: UserType<String> {

    override fun typeName(): String {
        return String::class.java.getTypeName()
    }

    override fun create(): String {
        return ""
    }

    override fun clone(o: String?): String? {
        return o
    }

    @Throws(IOException::class)
    override fun readValue(input: InputStreamReader): String {
        BufferedReader(input).use { br ->
            val line = br.readLine()
            return line?.trim { it <= ' ' } ?: ""
        }
    }

    override fun parseValue(string: String?): String {
        return string ?: ""
    }

    override fun getTypeComparator(): Comparator<String> {
        return Comparator { o1: Any?, o2: Any? ->
            if (o1 !is String || o2 !is String) {
                throw ClassCastException()
            }
            o1.compareTo(o2)
        }
    }
}